runtime/js-compute-runtime/builtins/cache-simple.cpp

```diff
@@ -156,6 +156,446 @@ JS::Result<std::string> createSurrogateKeyFromCacheKey(JSContext *cx,
 
 } // namespace
 
+bool SimpleCache::getOrSetThenHandler(JSContext *cx, JS::HandleObject owner, JS::HandleValue extra,
+                                      JS::CallArgs args) {
+  MOZ_ASSERT(extra.isObject());
+  JS::RootedObject extraObj(cx, &extra.toObject());
+  JS::RootedValue handleVal(cx);
+  JS::RootedValue promiseVal(cx);
+  if (!JS_GetProperty(cx, extraObj, "promise", &promiseVal)) {
+    return false;
+  }
+  MOZ_ASSERT(promiseVal.isObject());
+  JS::RootedObject promise(cx, &promiseVal.toObject());
+  if (!promise) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  if (!JS_GetProperty(cx, extraObj, "handle", &handleVal)) {
+    return RejectPromiseWithPendingError(cx, promise);
+  }
+  MOZ_ASSERT(handleVal.isInt32());
+  fastly_cache_handle_t handle = handleVal.toInt32();
+  fastly_error_t err;
+  JS::RootedValue keyVal(cx);
+  if (!JS_GetProperty(cx, extraObj, "key", &keyVal)) {
+    if (!fastly_transaction_cancel(handle, &err)) {
```

Comment:
We could come up with an RAII transaction class that would automatically cancel the transaction, unless explicitly released. No need to do that in this PR, but it would be a nice cleanup :)